chanfitness/chanfitness.go

```diff
@@ -0,0 +1,226 @@
+// to start the monitoring process.
+func NewScoreStore(config *Config) (*ScoreStore, error) {
+	scorer := &ScoreStore{
+		cfg:      config,
+		channels: make(map[uint64]*ChannelScore),
+		quit:     make(chan struct{}),
+	}
+
+	return scorer, nil
+}
+
+// Start adds all existing open channels to the score store and starts monitoring
+// for channel opened/closed events so that it can add and remove channels as
+// appropriate.
+func (c *ScoreStore) Start() error {
+	// Add the existing set of channels to the score store. This is required
+	// because channel events will not be triggered for channels that exist
+	// at startup time.
+	channels, err := c.cfg.GetOpenChannels()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Adding %v channels to score store", len(channels))
+
+	for _, ch := range channels {
+		var peerKey [33]byte
+		copy(peerKey[:], ch.IdentityPub.SerializeCompressed())
+
+		c.OpenChannel(ch.ShortChanID().ToUint64(), peerKey)
+	}
+
+	// channelUpdates provides a subscription to channel events which is used to
+	// listen for the creation of new channels and closing of existing channels.
+	channelUpdates, err := c.cfg.ChannelEvents()
+	if err != nil {
+		return err
+	}
+
+	go c.MonitorChannelEvents(
+		channelUpdates.Updates(),
+		channelUpdates.Cancel,
+	)
+
+	return nil
+}
+
+// Stop terminates all goroutines running for individual channels being monitored
+// and all scorer level goroutines which are running.
+func (c *ScoreStore) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	log.Info("Stopping score store")
+
+	// Stop the MonitorPeer goroutines running for each channel.
+	for _, ch := range c.channels {
+		ch.Stop()
+	}
+
+	// Stop the MonitorChannelEvents goroutine.
+	c.quit <- struct{}{}
+}
+
+// GetScore returns the score for the given channel id, and bool indicating
+// whether the channel is known to the score store.
+func (c *ScoreStore) GetScore(channelID uint64) (*ChannelScore, bool) {
+	c.Lock()
+	defer c.Unlock()
+
+	score, ok := c.channels[channelID]
+	return score, ok
+}
+
+// OpenChannel adds new channels to the ScoreStore's map of channels being
+// monitored and records a channel opened event. If the channel is already
+// present in the map, it is not overwritten and a channel opened event is not
+// recorded. If the channel is not being monitored, it starts a monitoring
+// goroutine to track when the channel's remote peer is online.
+func (c *ScoreStore) OpenChannel(channelID uint64, peer [33]byte) {
+	c.Lock()
+	defer c.Unlock()
+
+	score, ok := c.channels[channelID]
+	if !ok {
+		// The channel is not already known to the scorer,
+		// so must be initialized.
+		score = &ChannelScore{
+			id:        channelID,
+			peer:      peer,
+			quit:      make(chan struct{}, 1),
+			firstSeen: time.Now(),
+		}
+
+		// Since this is the first time this channel has been seen, a
+		// channel opened event is recorded.
+		score.RecordEvent(channelOpenedEvent)
+	}
+
+	if !score.monitoring {
```

Comment:
Seems like the goroutine can just be moved into the `!ok` scope, and `monitoring` removed?